File: Old/tests/struct_comparison_test.rs
use rumina::ast::BinOp;
use rumina::interpreter::Interpreter;
use rumina::value::Value;
use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;

#[test]
fn test_struct_not_equal_null() {
    let mut interpreter = Interpreter::new();
    let mut map = HashMap::new();
    map.insert("x".to_string(), Value::Int(10));
    let struct_val = Value::Struct(Rc::new(RefCell::new(map)));
    let null_val = Value::Null;

    // struct != null should be true
    let result = interpreter.eval_binary_op(&struct_val, BinOp::NotEqual, &null_val);
    assert!(result.is_ok(), "struct != null should be supported");
    assert_eq!(result.unwrap(), Value::Bool(true));
}

#[test]
fn test_struct_equal_null() {
    let mut interpreter = Interpreter::new();
    let mut map = HashMap::new();
    map.insert("x".to_string(), Value::Int(10));
    let struct_val = Value::Struct(Rc::new(RefCell::new(map)));
    let null_val = Value::Null;

    // struct == null should be false
    let result = interpreter.eval_binary_op(&struct_val, BinOp::Equal, &null_val);
    assert!(result.is_ok(), "struct == null should be supported");
    assert_eq!(result.unwrap(), Value::Bool(false));
}

#[test]
fn test_struct_equal_same_reference() {
    let mut interpreter = Interpreter::new();
    let mut map = HashMap::new();
    map.insert("x".to_string(), Value::Int(10));
    let struct_val = Value::Struct(Rc::new(RefCell::new(map)));
    let struct_val2 = struct_val.clone();

    // Same reference should be equal
    let result = interpreter.eval_binary_op(&struct_val, BinOp::Equal, &struct_val2);
    assert!(
        result.is_ok(),
        "struct == struct (same ref) should be supported"
    );
    assert_eq!(result.unwrap(), Value::Bool(true));
}

#[test]
fn test_struct_equal_different_reference() {
    let mut interpreter = Interpreter::new();
    let mut map1 = HashMap::new();
    map1.insert("x".to_string(), Value::Int(10));
    let struct_val1 = Value::Struct(Rc::new(RefCell::new(map1)));

    let mut map2 = HashMap::new();
    map2.insert("x".to_string(), Value::Int(10));
    let struct_val2 = Value::Struct(Rc::new(RefCell::new(map2)));

    // Different references should not be equal (even with same content)
    let result = interpreter.eval_binary_op(&struct_val1, BinOp::Equal, &struct_val2);
    assert!(
        result.is_ok(),
        "struct == struct (diff ref) should be supported"
    );
    assert_eq!(result.unwrap(), Value::Bool(false));
}

#[test]
fn test_struct_not_equal_different_reference() {
    let mut interpreter = Interpreter::new();
    let mut map1 = HashMap::new();
    map1.insert("x".to_string(), Value::Int(10));
    let struct_val1 = Value::Struct(Rc::new(RefCell::new(map1)));

    let mut map2 = HashMap::new();
    map2.insert("x".to_string(), Value::Int(10));
    let struct_val2 = Value::Struct(Rc::new(RefCell::new(map2)));

    // Different references should be not equal
    let result = interpreter.eval_binary_op(&struct_val1, BinOp::NotEqual, &struct_val2);
    assert!(
        result.is_ok(),
        "struct != struct (diff ref) should be supported"
    );
    assert_eq!(result.unwrap(), Value::Bool(true));
}

#[test]
fn test_struct_not_equal_other_types() {
    let mut interpreter = Interpreter::new();
    let mut map = HashMap::new();
    map.insert("x".to_string(), Value::Int(10));
    let struct_val = Value::Struct(Rc::new(RefCell::new(map)));

    // Test with int
    let int_val = Value::Int(10);
    let result = interpreter.eval_binary_op(&struct_val, BinOp::NotEqual, &int_val);
    assert!(result.is_ok(), "struct != int should be supported");
    assert_eq!(result.unwrap(), Value::Bool(true));

    // Test with string
    let string_val = Value::String("test".to_string());
    let result = interpreter.eval_binary_op(&struct_val, BinOp::NotEqual, &string_val);
    assert!(result.is_ok(), "struct != string should be supported");
    assert_eq!(result.unwrap(), Value::Bool(true));

    // Test with bool
    let bool_val = Value::Bool(true);
    let result = interpreter.eval_binary_op(&struct_val, BinOp::NotEqual, &bool_val);
    assert!(result.is_ok(), "struct != bool should be supported");
    assert_eq!(result.unwrap(), Value::Bool(true));
}

#[test]
fn test_struct_equal_other_types() {
    let mut interpreter = Interpreter::new();
    let mut map = HashMap::new();
    map.insert("x".to_string(), Value::Int(10));
    let struct_val = Value::Struct(Rc::new(RefCell::new(map)));

    // Test with int
    let int_val = Value::Int(10);
    let result = interpreter.eval_binary_op(&struct_val, BinOp::Equal, &int_val);
    assert!(result.is_ok(), "struct == int should be supported");
    assert_eq!(result.unwrap(), Value::Bool(false));

    // Test with string
    let string_val = Value::String("test".to_string());
    let result = interpreter.eval_binary_op(&struct_val, BinOp::Equal, &string_val);
    assert!(result.is_ok(), "struct == string should be supported");
    assert_eq!(result.unwrap(), Value::Bool(false));
}

#[test]
fn test_null_not_equal_struct() {
    let mut interpreter = Interpreter::new();
    let mut map = HashMap::new();
    map.insert("x".to_string(), Value::Int(10));
    let struct_val = Value::Struct(Rc::new(RefCell::new(map)));
    let null_val = Value::Null;

    // null != struct should be true (reversed order)
    let result = interpreter.eval_binary_op(&null_val, BinOp::NotEqual, &struct_val);
    assert!(result.is_ok(), "null != struct should be supported");
    assert_eq!(result.unwrap(), Value::Bool(true));
}

File: CRumina/tests/struct_comparison_test.cc
#include <test_framework.h>
#include <interpreter.h>
#include <value.h>
#include <ast.h>
#include <memory>
#include <unordered_map>

using namespace rumina;
using namespace rumina::test;

std::shared_ptr<std::unordered_map<std::string, Value>> create_test_struct() {
    auto map = std::make_shared<std::unordered_map<std::string, Value>>();
    (*map)["x"] = Value(static_cast<int64_t>(10));
    return map;
}

void test_struct_not_equal_null() {
    Interpreter interpreter;
    auto struct_val = Value::makeStruct(create_test_struct());
    Value null_val;

    auto result = interpreter.eval_binary_op(struct_val, BinOp::NotEqual, null_val);
    assert_ok(result);
    auto value = result.value();
    assert_eq(value.getType(), Value::Type::Bool);
    assert_true(value.getBool());
}

void test_struct_equal_null() {
    Interpreter interpreter;
    auto struct_val = Value::makeStruct(create_test_struct());
    Value null_val;

    auto result = interpreter.eval_binary_op(struct_val, BinOp::Equal, null_val);
    assert_ok(result);
    auto value = result.value();
    assert_eq(value.getType(), Value::Type::Bool);
    assert_false(value.getBool());
}

void test_struct_equal_same_reference() {
    Interpreter interpreter;
    auto map = create_test_struct();
    auto struct_val1 = Value::makeStruct(map);
    auto struct_val2 = Value::makeStruct(map);  // Same underlying map

    auto result = interpreter.eval_binary_op(struct_val1, BinOp::Equal, struct_val2);
    assert_ok(result);
    auto value = result.value();
    assert_eq(value.getType(), Value::Type::Bool);
    assert_true(value.getBool());
}

void test_struct_equal_different_reference() {
    Interpreter interpreter;
    auto struct_val1 = Value::makeStruct(create_test_struct());
    auto struct_val2 = Value::makeStruct(create_test_struct());  // Different map

    auto result = interpreter.eval_binary_op(struct_val1, BinOp::Equal, struct_val2);
    assert_ok(result);
    auto value = result.value();
    assert_eq(value.getType(), Value::Type::Bool);
    assert_false(value.getBool());
}

void test_struct_not_equal_different_reference() {
    Interpreter interpreter;
    auto struct_val1 = Value::makeStruct(create_test_struct());
    auto struct_val2 = Value::makeStruct(create_test_struct());

    auto result = interpreter.eval_binary_op(struct_val1, BinOp::NotEqual, struct_val2);
    assert_ok(result);
    auto value = result.value();
    assert_eq(value.getType(), Value::Type::Bool);
    assert_true(value.getBool());
}

void test_struct_not_equal_other_types() {
    Interpreter interpreter;
    auto struct_val = Value::makeStruct(create_test_struct());

    Value int_val(static_cast<int64_t>(10));
    auto result = interpreter.eval_binary_op(struct_val, BinOp::NotEqual, int_val);
    assert_ok(result);
    auto value = result.value();
    assert_eq(value.getType(), Value::Type::Bool);
    assert_true(value.getBool());

    Value string_val("test");
    result = interpreter.eval_binary_op(struct_val, BinOp::NotEqual, string_val);
    assert_ok(result);
    value = result.value();
    assert_eq(value.getType(), Value::Type::Bool);
    assert_true(value.getBool());

    Value bool_val(true);
    result = interpreter.eval_binary_op(struct_val, BinOp::NotEqual, bool_val);
    assert_ok(result);
    value = result.value();
    assert_eq(value.getType(), Value::Type::Bool);
    assert_true(value.getBool());
}

void test_struct_equal_other_types() {
    Interpreter interpreter;
    auto struct_val = Value::makeStruct(create_test_struct());

    Value int_val(static_cast<int64_t>(10));
    auto result = interpreter.eval_binary_op(struct_val, BinOp::Equal, int_val);
    assert_ok(result);
    auto value = result.value();
    assert_eq(value.getType(), Value::Type::Bool);
    assert_false(value.getBool());

    Value string_val("test");
    result = interpreter.eval_binary_op(struct_val, BinOp::Equal, string_val);
    assert_ok(result);
    value = result.value();
    assert_eq(value.getType(), Value::Type::Bool);
    assert_false(value.getBool());
}

void test_null_not_equal_struct() {
    Interpreter interpreter;
    Value null_val;
    auto struct_val = Value::makeStruct(create_test_struct());

    auto result = interpreter.eval_binary_op(null_val, BinOp::NotEqual, struct_val);
    assert_ok(result);
    auto value = result.value();
    assert_eq(value.getType(), Value::Type::Bool);
    assert_true(value.getBool());
}

int main() {
    TestRunner runner;
    
    runner.add_test("struct_not_equal_null", test_struct_not_equal_null);
    runner.add_test("struct_equal_null", test_struct_equal_null);
    runner.add_test("struct_equal_same_reference", test_struct_equal_same_reference);
    runner.add_test("struct_equal_different_reference", test_struct_equal_different_reference);
    runner.add_test("struct_not_equal_different_reference", test_struct_not_equal_different_reference);
    runner.add_test("struct_not_equal_other_types", test_struct_not_equal_other_types);
    runner.add_test("struct_equal_other_types", test_struct_equal_other_types);
    runner.add_test("null_not_equal_struct", test_null_not_equal_struct);
    
    return runner.run_all();
}
